Code/Engine/RendererCore/Material/Implementation/MaterialManager.cpp

```diff
@@ -0,0 +1,694 @@
+}
+
+void ezMaterialManager::RegisterMaterial(ezMaterialResource* pMaterial)
+{
+  EZ_PROFILE_SCOPE("RegisterMaterial");
+  EZ_LOCK(m_MaterialShaderMutex);
+  EZ_LOCK(m_ExtractionMutex);
+
+  const ezShaderResourceHandle hOldShader = pMaterial->m_hShader;
+  const ezMaterialResource::ezMaterialId oldId = pMaterial->m_MaterialId;
+  const bool bShaderChanged = pMaterial->m_mDesc.m_hShader != hOldShader;
+
+  if (!hOldShader.IsValid())
+  {
+    if (pMaterial->m_mDesc.m_hShader.IsValid())
+    {
+      // Newly created material
+      pMaterial->m_hShader = pMaterial->m_mDesc.m_hShader;
+      MaterialShaderConstants& msc = GetShaderConstants(pMaterial->m_hShader);
+      pMaterial->m_MaterialId = msc.AddMaterial(pMaterial->GetResourceHandle());
+    }
+  }
+  else if (bShaderChanged)
+  {
+    // Material reloaded and changed shader
+    UnregisterMaterial(pMaterial);
+
+    // Create new material registration
+    if (pMaterial->m_mDesc.m_hShader.IsValid())
+    {
+      pMaterial->m_hShader = pMaterial->m_mDesc.m_hShader;
+      MaterialShaderConstants& msc = GetShaderConstants(pMaterial->m_hShader);
+      pMaterial->m_MaterialId = msc.AddMaterial(pMaterial->GetResourceHandle());
+    }
+    else
+    {
+      pMaterial->m_hShader = pMaterial->m_mDesc.m_hShader;
+      pMaterial->m_MaterialId.Invalidate();
+    }
+
+    // Fire event that the shader / id has changed and needs to be invalidated.
+    ezMaterialShaderChanged change;
+    change.m_hMaterial = pMaterial->GetResourceHandle();
+    change.m_hOldShader = hOldShader;
+    change.m_OldId = oldId;
+    change.m_hNewShader = pMaterial->m_hShader;
+    change.m_NewId = pMaterial->m_MaterialId;
+    s_MaterialShaderChangedEvent.Broadcast(change);
+  }
+  else
+  {
+    // Material reloaded with same shader. Nothing to do except marking material dirty below.
+  }
+  m_ModifiedMaterials.Insert(pMaterial->GetResourceHandle());
+}
+
+void ezMaterialManager::UnregisterMaterial(ezMaterialResource* pMaterial)
+{
+  EZ_LOCK(m_MaterialShaderMutex);
+  EZ_LOCK(m_ExtractionMutex);
+
+  if (pMaterial->m_hShader.IsValid())
+  {
+    MaterialShaderConstants& msc = GetShaderConstants(pMaterial->m_hShader);
+    msc.RemoveMaterial(pMaterial->m_MaterialId);
+    m_RemovedMaterials.PushBack(pMaterial);
+  }
+}
+
+void ezMaterialManager::ExtractMaterial(ezMaterialResource* pMaterial, ezMaterialManager::ExtractedMaterial& extractedMaterial)
+{
+  EZ_PROFILE_SCOPE("ExtractMaterial");
+  extractedMaterial.m_hMaterial = pMaterial->GetResourceHandle();
+  extractedMaterial.m_hShader = pMaterial->m_mDesc.m_hShader;
+  extractedMaterial.m_MaterialId = pMaterial->m_MaterialId;
+  extractedMaterial.m_DirtyFlags = pMaterial->m_DirtyFlags;
+  for (ezMaterialResource::DirtyFlags::Enum flag : pMaterial->m_DirtyFlags)
+  {
+    switch (flag)
+    {
+      case ezMaterialResource::DirtyFlags::Parameter:
+        extractedMaterial.m_Parameters = std::move(pMaterial->m_mDesc.m_Parameters);
```

Comment:
Are you sure about these moves here? The material desc is "dead" afterwards. I think we should copy here but move in ApplyMaterialChanges?

Reply:
Woops! Thanks for noticing! Fixed and updated the renderer test to detect this.